test/remote/gateways/remote_shift4_v2_test.rb

```diff
@@ -0,0 +1,341 @@
+require 'test_helper'
+
+class RemoteShift4V2Test < Test::Unit::TestCase
+  CHARGE_ID_REGEX = /char_[a-zA-Z\d]+/
+  TOKEN_ID_REGEX  = /tok_[a-zA-Z\d]+/
+
+  def setup
+    @gateway = Shift4V2Gateway.new(fixtures(:shift4_v2))
+
+    @amount = 2000
+    @refund_amount = 300
+    @credit_card = credit_card('4242424242424242')
+    @declined_card = credit_card('4916018475814056')
+    @new_credit_card = credit_card('4012888888881881')
+    @invalid_token = 'tok_invalid'
+
+    @options = {
+      description: 'ActiveMerchant test charge',
+      email: 'foo@example.com'
+    }
+  end
+
+  def test_successful_store
+    response = @gateway.store(@credit_card, @options)
+    assert_success response
+    assert_match %r(^cust_\w+$), response.authorization
+    assert_equal 'customer', response.params['objectType']
+    assert_match %r(^card_\w+$), response.params['cards'][0]['id']
+    assert_equal 'card', response.params['cards'][0]['objectType']
+
+    @options[:customer_id] = response.authorization
+    response = @gateway.store(@new_credit_card, @options)
+    assert_success response
+    assert_match %r(^card_\w+$), response.params['card']['id']
+    assert_equal @options[:customer_id], response.params['card']['customerId']
+
+    response = @gateway.customer(@options)
+    assert_success response
+    assert_equal @options[:customer_id], response.params['id']
+    assert_equal '401288', response.params['cards'][0]['first6']
+    assert_equal '1881', response.params['cards'][0]['last4']
+    assert_equal '424242', response.params['cards'][1]['first6']
+    assert_equal '4242', response.params['cards'][1]['last4']
+  end
+
+  # def test_dump_transcript
+  #   skip("Transcript scrubbing for this gateway has been tested.")
+  #   dump_transcript_and_fail(@gateway, @amount, @credit_card, @options)
+  # end
+
+  def test_transcript_scrubbing
+    transcript = capture_transcript(@gateway) do
+      @gateway.purchase(@amount, @credit_card, @options)
+    end
+    transcript = @gateway.scrub(transcript)
+    assert_scrubbed(@credit_card.number, transcript)
+    assert_scrubbed(@credit_card.verification_value, transcript)
+    assert_scrubbed(@credit_card.year, transcript)
+    assert_scrubbed(@gateway.options[:secret_key], transcript)
+  end
+
+  def test_successful_purchase
+    response = @gateway.purchase(@amount, @credit_card, @options)
+    assert_success response
+    assert_equal 'Transaction approved', response.message
+    assert_equal 'foo@example.com', response.params['metadata']['email']
+    assert_match CHARGE_ID_REGEX, response.authorization
+  end
+
+  def test_successful_purchase_with_three_ds_data
+    @options[:three_d_secure] = {
+      version: '2.1.0',
+      eci: '05',
+      cavv: '3q2+78r+ur7erb7vyv66vv////8=',
+      acs_transaction_id: '6546464645623455665165+qe-jmhabcdefg',
+      authentication_response_status: 'Y'
+    }
```

Comment:
The `@options[:three_d_secure]` is only used here but it could be moved to a variable in the `setup` to make it more readable, also the version in this object is `2.1.0` and in the unit tests it is `1.0.2`, is this correct?